### PR TITLE
Model relationships

### PR DESCRIPTION
**Pull request by markdamobi**
- Changed some existing migrations.I know existing migrations shouldn't really be changed. but I don't think anyone has done much on this repo yet so it's okay in this case. If this gets merged though, subsequent changes to db structure should be with new migrations.  
- Added some new migrations. 
- Setup relationships. 
- add seed data for orgs and org types.

### CODE REVIEW

File: db/migrate/20190202210545_create_donations.rb [hunk 1]
```diff
@@ -1,7 +1,9 @@
 class CreateDonations < ActiveRecord::Migration[5.2]
   def change
     create_table :donations do |t|
-
+      t.references :user, foreign_key: true
+      t.references :organization, foreign_key: true
+      t.integer :amount
```
Comment (sanghun89) [inline db/migrate/20190202210545_create_donations.rb]: It's literally the beginning of the codebase, but as a rule of thumb, you should never edit existing migration files when working with rails db migration files. You should always generate migration files, say `AddReferencesToDonations` and/or `AddAmoundFieldToDonations`.

The better practice is to generate migration files at all times. This is so that no devs contributing to this project will be out-of-sync when migrating or rolling back database schema.

This applies to all your migration edits

File: app/models/user.rb [hunk 2]
```diff
@@ -1,3 +1,7 @@
 class User < ApplicationRecord
-  self.primary_key = :id
+  has_many :user_fav_orgs, dependent: :destroy
```
Comment (sanghun89) [inline app/models/user.rb]: This seems a little too over-architected in my opinion. Can we not refer this to `Organization` class so you don't have to create another model for it? 

Also this will make it harder to have unique favorite organizations

File: db/seeds.rb [hunk 3]
```diff
@@ -5,3 +5,9 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+
+org_types = ['community_development', 'education', 'religion', 'animals', 'health', 'human_rights', 'human_services', 'international', 'environment']
```
Comment (sanghun89) [inline db/seeds.rb]: So at the moment the old application keeps this as the source of truth for all existing organization types. I personally don't think we should go with the approach of creating tables for type. For me the reasons are:

1. You are only going to have a finite number of these types
2. It is difficult in code to distinguish one type from another.

I personally think these should be something like https://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html
is better to know these types in compile time, and not relying on a seed file to learn the source of truth of types.

File: db/migrate/20190202204545_create_users.rb [hunk 4]
```diff
@@ -1,10 +1,23 @@
 class CreateUsers < ActiveRecord::Migration[5.2]
   def change
-    create_table :users, id: false do |t|
-      t.string :id
-      t.string :name
+    create_table :users do |t|
+      t.string :name, null: false
+      t.string :nick_name
+      t.string :fb_id, null: false
```
Comment (sanghun89) [inline db/migrate/20190202204545_create_users.rb]: There was an issue commenting about how we want to keep the user flexible. We shouldn't enforce an `fb_id` if that's the case

File: app/models/organization.rb [hunk 5]
```diff
@@ -1,2 +1,5 @@
 class Organization < ApplicationRecord
+  has_many :user_fav_orgs, dependent: :destroy
+  has_many :donations, dependent: :destroy
+  belongs_to :organization_type, optional: true
```
Comment (sanghun89) [inline app/models/organization.rb]: I mentioned this in another review comment, but I think these should be enums.

File: app/models/organization.rb [hunk 6]
```diff
@@ -1,2 +1,5 @@
 class Organization < ApplicationRecord
+  has_many :user_fav_orgs, dependent: :destroy
```
Comment (sanghun89) [inline app/models/organization.rb]: Does an organization need to know that it belongs to a certain user's favorite org list?? I would think not

File: Gemfile [hunk 7]
```diff
@@ -46,8 +46,8 @@ group :development do
   gem 'web-console', '>= 3.3.0'
   gem 'listen', '>= 3.0.5', '< 3.2'
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
-  gem 'spring'
-  gem 'spring-watcher-listen', '~> 2.0.0'
+  # gem 'spring'
+  # gem 'spring-watcher-listen', '~> 2.0.0'
```
Comment (sanghun89) [inline Gemfile]: reason for this?

File: db/migrate/20190202210545_create_donations.rb (hunk 1 above)
Comment (markdamobi) [reply db/migrate/20190202210545_create_donations.rb]: Yup. totally agree. That's pretty much what I said in the first point above in my description.

File: db/migrate/20190202204545_create_users.rb (hunk 4 above)
Comment (markdamobi) [reply db/migrate/20190202204545_create_users.rb]: no problem. I can get rid of the constraint. I originally just thought the only point of entry to the site was through facebook login.

File: Gemfile (hunk 7 above)
Comment (markdamobi) [reply Gemfile]: I've run into some weird debugging issues in the past where some changes weren't being properly recognized with the way spring preloads the app sometimes. There were some instances, where I exist rails console but spring is still running. So i've had to manually kill some spring processes because of that.

**Comment by markdamobi** [comment]
Oh yeah, this is just exploration. If we decide to move forward with it, we can include rspec and I can write some unit tests.

File: app/models/organization.rb (hunk 6 above)
Comment (markdamobi) [reply app/models/organization.rb]: nah, I can get rid of this.

File: db/migrate/20190202204545_create_users.rb (hunk 4 above)
Comment (sanghun89) [reply db/migrate/20190202204545_create_users.rb]: Yea the design does enforce it. It was more from this comment from the issue 

> And my 2 cents for no. 2 would be to store the FB id separately from the user id. If ever want to add other auth services (like google or just email) using the FB id could pose a conflict.

I'd rather just keep this null and have the custom validation in rails instead of baking in the the constraint in postgres if the talk is about having authenticated source being more flexible.

File: app/models/user.rb (hunk 2 above)
Comment (markdamobi) [reply app/models/user.rb]: What do you mean by "unique" favorite organization? The fav_orgs column on the user from the data is a list. So a user has multiple favorite organizations. I thought it'd be easier to represent this using a join table because it seems like a many-to-many relationship. What did you have in mind?

File: db/seeds.rb (hunk 3 above)
Comment (markdamobi) [reply db/seeds.rb]: I thought about the enum types, but decided that it may cause some headache if I was trying to debug something by looking directly in the database since the enum types would be stored as integers. 
But I can definitely change it though if we don't want the extra table.

Comment (sanghun89) [reply db/seeds.rb]: I might be wrong, but I believe you can set it to where you store enum as strings

Comment (markdamobi) [reply db/seeds.rb]: Oh that'll be nice. I'll definitely look into that.

**Comment by markdamobi** [comment]
I just updated the pr to use enum. I'm using a shared enum between the User and Org models.. However, it's only on the application level though. I looked into an alternative [https://edgeguides.rubyonrails.org/active_record_postgresql.html#enumerated-types](here) to get it to the db level. It works, but then the schema.rb gets messed up because afaik it doesn't yet support custom postgres types. We can try to switch over to structure.sql instead of schema.rb, I don't think we should do that.
I think it's fine to have the enum just on the model level for now.

File: app/models/user.rb [hunk 8]
```diff
@@ -1,3 +1,10 @@
...
-  self.primary_key = :id
+  enum favorite_cause: Organization::org_types
+  
+  has_many :user_fav_orgs, dependent: :destroy
+  has_many :favorite_organizations, through: :user_fav_orgs, source: :organization
+  has_many :donations, dependent: :destroy
+
+  validates_presence_of :fb_id, presence: true #this can be removed when some other signup/login method is added. 
 end
+
```
Comment (arincon9) [inline app/models/user.rb]: We should remove this extra line.

File: app/models/user.rb (hunk 2 above)
Comment (sanghun89) [reply app/models/user.rb]: I mean simply put I thought having a property with array of ids would suffice, not a whole many-to-many association.

That being said, if you do want to stick with this, you have to mark **both** `user_id` and `organization_id` in `UserFavOrg` as primary keys. If you leave it as is, an user can potentially have multiple instances of `user_fav_org` pointing to the same org id.

Also can we alias this to just `fav_orgs` or spell the whole thing out (`favorite_orgs)`??

**Comment by sanghun89** [comment]
Hey @markdamobi, I know this was more an exploration, but being that the remaining issue kind of depend on this PR, do you think you can wrap up this PR where it's mergeable?

It wouldn't be much - I believe few things are:
- Addressing any outstanding feedbacks
- Create unit tests for these models and test associations (testing out enum would be nice as well)

Once it's in, I'll make an ultimate PR back to Great Heart's repository.

File: app/models/user.rb (hunk 2 above)
Comment (uuunk) [reply app/models/user.rb]: If you're going to have an array of ids you might as well make the association. You get all the rails/activerecord magic for free, so might as well take advantage.  If you do ids, you'd have to manually make the queries. (e.g. `User.first.favorite_organizations` gets you an (activerecord) array of orgs where as you'd have to do `Orgs.where(id: User.first.favorite_organizations)` if you just did ids.

Comment (uuunk) [reply app/models/user.rb]: I would agree to change the name to `user_favorite_organizations`, prob too much Java but i'm always a fan of fully naming things

Comment (uuunk) [reply app/models/user.rb]: And just noticed that "description" is a field on the user_fav_orgs table which is needed, so and id array would really be inadequate at that point. So now the join table is more than a "nice to have"

File: db/seeds.rb (hunk 3 above)
Comment (uuunk) [reply db/seeds.rb]: Yeah we do enums as strings using a gem, I think @sanghun89 posted one in slack. It's nice, best of both worlds IMO.

File: db/seed_files/orgs.yml [hunk 9]
```diff
@@ -0,0 +1,121 @@
+---
+- :name: PBS
+  :url: https://www.facebook.com/pbs
```
Comment (uuunk) [inline db/seed_files/orgs.yml]: could we change these to `fb_url:` in case we ever want to seed the actual url in addition?

File: db/seeds.rb [hunk 10]
```diff
@@ -5,3 +5,9 @@
...
+org_types.each{|org_type| OrganizationType.create!(name: org_type)} 
+
+orgs = YAML::load_file(Rails.root.join("db/seed_files/orgs.yml"))
+orgs.each{|org| Organization.create!(name: org[:name], fb_url: org[:url])}
```
Comment (uuunk) [inline db/seeds.rb]: If the seed were run more than once, would this blow up? if we're not sure, would be worth double checking. The seed file should be able to run any number of times and still end up with the same consistent result and never fail for data reasons.

File: app/models/user.rb (hunk 2 above)
Comment (uuunk) [reply app/models/user.rb]: > * Create unit tests for these models and test associations (testing out enum would be nice as well)

I'm not too worried about unit tests for the associations, at least I don't think it's worth blocking a PR for. I'm happy to acquiesce if we think that's a slippery slope though.

**Comment by uuunk** [comment]
> * Create unit tests for these models and test associations (testing out enum would be nice as well)

I'm not too worried about unit tests for the associations, at least not enough to hold up a PR for. But I'm happy to acquiesce if we think that's a slippery slope.

File: app/models/organization.rb [hunk 11]
```diff
@@ -1,2 +1,15 @@
 class Organization < ApplicationRecord
+  enum org_type: {
+    cummunity_development: 'community_development',
```
Comment (arincon9) [inline app/models/organization.rb]: Typo on the key `community_development`

File: app/models/user.rb (hunk 2 above)
Comment (sanghun89) [reply app/models/user.rb]: Ah I see, I didn't see the `description` part, but I guess that raises the question of why is `description` on each organization list?

And correct me if I'm wrong, but I think you would need composite primary keys on `user_id` and `organization_id` on that table correct?

File: app/models/organization.rb (hunk 11 above)
Comment (markdamobi) [reply app/models/organization.rb]: good catch!

File: db/seeds.rb (hunk 10 above)
Comment (markdamobi) [reply db/seeds.rb]: yeah, good point. Right now it'll blow up because because there's uniqueness validation on both fields, so there won't be any corrupted data. But I'll give it better handling so it runs more smoothly.

File: db/seed_files/orgs.yml (hunk 9 above)
Comment (markdamobi) [reply db/seed_files/orgs.yml]: no problem.

File: app/models/user.rb (hunk 2 above)
Comment (markdamobi) [reply app/models/user.rb]: - I put the description field there because I saw `fav_orgs_text` on the user model in the data. 
- What would be an advantage of having a composite primary key? I was just going to leave the primary key as the id rails generate. But then I'll add uniqueness validation on the 2 columns. That should prevent multiple instances.

**Comment by sanghun89** [comment]
@markdamobi 

> > * Create unit tests for these models and test associations (testing out enum would be nice as well)
> 
> I'm not too worried about unit tests for the associations, at least not enough to hold up a PR for. But I'm happy to acquiesce if we think that's a slippery slope.

I'd vote for adding it - although it seems small it would set the ground work for more stories to come (I'm also a stickler for testing boundaries). This PR was estimated at a 5, and having it this far in one day is awesome - I feel like we can squeeze a little more time to set the foundation of this a little stronger

File: app/models/user.rb (hunk 2 above)
Comment (sanghun89) [reply app/models/user.rb]: The primary composite key was to prevent exactly that, "multiple instances."

Comment (sanghun89) [reply app/models/user.rb]: but yeah grouping both of them as unique would work as well

**Comment by markdamobi** [comment]
> > * Create unit tests for these models and test associations (testing out enum would be nice as well)
> 
> I'm not too worried about unit tests for the associations, at least not enough to hold up a PR for. But I'm happy to acquiesce if we think that's a slippery slope.

We can get testing setup later, but not in this pr so it won't hold anyone back that's trying to do something else.

File: app/models/user.rb (hunk 2 above)
Comment (sanghun89) [reply app/models/user.rb]: We should also try to refactor names when we can and seems appropriate, and not just blindly follow what the previous application did

**Comment by markdamobi** [comment]
> I'd vote for adding it - although it seems small it would set the ground work for more stories to come (I'm also a stickler for testing boundaries). This PR was estimated at a 5, and having it this far in one day is awesome - I feel like we can squeeze a little more time to set the foundation of this a little stronger

Good point. I don't know if I'll be able to get to it though. but just in case I can, would you mind writing a list of things to test? Not everything, but just a good starting point and then other tests can be added in later.

**Comment by sanghun89** [comment]
> > I'd vote for adding it - although it seems small it would set the ground work for more stories to come (I'm also a stickler for testing boundaries). This PR was estimated at a 5, and having it this far in one day is awesome - I feel like we can squeeze a little more time to set the foundation of this a little stronger
> 
> Good point. I don't know if I'll be able to get to it though. but just in case I can, would you mind writing a list of things to test? Not everything, but just a good starting point and then other tests can be added in later.

I'd expect to be just 4 rspec files, reflecting your models. Testing of associations are really simple, you can use https://github.com/thoughtbot/shoulda to ensure your associations. This way if someone was to change it down the road, you don't have weird issues. 

That being said, if you want to stop at just minor changes, could you create a trello issue on addressing unit tests? I can pick it up later after work and finish that up, and I can make a PR to great heart's repo as soon as you're done

File: app/models/user.rb (hunk 2 above)
Comment (markdamobi) [reply app/models/user.rb]: yeah, I agree. That's why I used favorite_cause_description instead of fav_cause_text on the user. I can make more changes if there are others that need to be changes.

**Comment by markdamobi** [comment]
I think I've addressed most of the feedback(except for the testing). Let me know if I missed something.
